### PR TITLE
fix(hicache): bound L3 prefetch prefix by sidecar coverage

### LINKED ISSUE #30057: [Bug] HiCache L3 partial prefetch can restore DSA/NSA KV pages without matching INDEXER sidecar pages
### Checklist

- [x] I searched related issues but found no exact solution.
- [x] The relevant code path is present in the latest main snapshot I checked (`430418e218`).
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] This is a bug report, not a general question.
- [x] This issue is written in English.

### Describe the bug

DSA/NSA models have required sidecar state in addition to the normal KV pages, such as the `INDEXER` pool used by `DSATokenToKVPool`. When HiCache L3 prefetch publishes a partial KV prefix, the host/L2 cache can contain KV pages whose required `INDEXER` sidecar pages were not loaded. A later L2 load-back can then restore KV pages without the matching INDEXER state, which can produce incorrect output.

This is related to, but different from, #28895. That issue reports missing DSA draft/indexer state causing lower MTP acceptance after cache hits. This issue is about prefix consistency when L3 prefetch partially completes: the usable L2 prefix should be constrained by both KV pages and required sidecar pages.

Current code shape in main (`430418e218`) appears to allow this for partial prefetches:

1. `HiRadixCache.can_terminate_prefetch()` / `UnifiedRadixCache` only waits for `pool_transfers_done` when the KV prefetch is fully completed. With `timeout` or `best_effort`, a partial KV result can be published.
2. `HybridCacheController._page_transfer()` only fetches extra pools after KV fully completes:

```python
kv_completed_pages = operation.completed_tokens // self.page_size
if operation.pool_transfers and kv_completed_pages == len(operation.hash_value):
    self._sync_trailing_keys(...)
    self._resolve_sidecar_derived_pool_transfers(operation)
    results = self.storage_backend.batch_get_v2(operation.pool_transfers)
    operation.pool_storage_result.update_extra_pool_hit_pages(results)
operation.pool_transfers_done = True
```

If a 128K-token L3 query times out after only 16K KV tokens, `kv_completed_pages != len(operation.hash_value)`, so the INDEXER sidecar is not fetched. The 16K KV prefix may still be inserted into the host radix tree and can later be treated as an L2 hit.

For DSA/NSA, this is unsafe unless L2 load-back applies a prefix rule such as:

```text
usable_l2_tokens = min(kv_l2_tokens, required_sidecar_l2_tokens)
```

For `PoolHitPolicy.ALL_PAGES` sidecars like `INDEXER`, a missing sidecar page should truncate the usable prefix at the first missing page, not allow loading the corresponding KV pages alone.

### Reproduction

A minimal way to reproduce the state is to make INDEXER L3 `batch_get_v2` slower than KV L3 prefetch and use a policy that can publish before all requested pages are available.

Observed with a GLM-5.2 DSA/NSA-style model using HiCache L3 storage, `--hicache-storage-prefetch-policy best_effort`, and a long prompt where the second request hits L3. Internal host/IP/backend details are omitted, but the relevant flags are:

```bash
python3 -m sglang.launch_server \
  --model-path <glm-5.2-dsa-model> \
  --trust-remote-code \
  --enable-hierarchical-cache \
  --enable-cache-report \
  --tp 4 \
  --kv-cache-dtype fp8_e4m3 \
  --max-prefill-tokens 73728 \
  --max-output-length 131072 \
  --hicache-write-policy write_back \
  --hicache-io-backend kernel \
  --hicache-mem-layout page_first \
  --hicache-storage-backend <l3-storage-backend> \
  --hicache-storage-prefetch-policy best_effort \
  --attention-backend nsa \
  --nsa-prefill-backend trtllm \
  --nsa-decode-backend trtllm \
  --enable-nsa-prefill-context-parallel \
  --nsa-prefill-cp-mode round-robin-split \
  --attn-cp-size 4
```

The prompt length was about 128K input tokens. The request hit about 1985 L3 pages. When INDEXER sidecar loading was delayed, the prefetch publish logs showed different completed KV prefixes across ranks while sidecar transfer was not done:

```text
HiCache hybrid prefetch publishing req=<id> policy=best_effort completed_tokens=32768 queried_tokens=127040 pool_transfers_done=False pool_storage_result=PoolTransferResult(kv_hit_pages=1985, extra_pool_hit_pages={})
HiCache hybrid prefetch publishing req=<id> policy=best_effort completed_tokens=32768 queried_tokens=127040 pool_transfers_done=False pool_storage_result=PoolTransferResult(kv_hit_pages=1985, extra_pool_hit_pages={})
HiCache hybrid prefetch publishing req=<id> policy=best_effort completed_tokens=49152 queried_tokens=127040 pool_transfers_done=False pool_storage_result=PoolTransferResult(kv_hit_pages=1985, extra_pool_hit_pages={})
HiCache hybrid prefetch publishing req=<id> policy=best_effort completed_tokens=81920 queried_tokens=127040 pool_transfers_done=False pool_storage_result=PoolTransferResult(kv_hit_pages=1985, extra_pool_hit_pages={})
```

With the above state, the second request produced corrupted/garbled output. Switching to a policy/logging setup that effectively waited until more sidecar work had completed avoided the corruption in the same workload.

This is not limited to `best_effort`. With `timeout`, if the KV L3 prefetch publishes a partial prefix, the sidecar fetch is also skipped because KV did not fully complete. With `wait_complete`, the already existing `pool_transfers_done` guard helps only once KV fully completes; it does not address the general invariant that L2 application must never use a KV prefix longer than the required sidecar prefix.

### Expected behavior

For hybrid HiCache pools with required sidecar pages, especially DSA/NSA `INDEXER` with `PoolHitPolicy.ALL_PAGES`:

- L3 prefetch may still be best-effort or timeout-based.
- A request should not block indefinitely on slow sidecar IO.
- However, when applying/restoring L2 cache to GPU, the usable prefix should be the minimum safe prefix across KV and required sidecar pools.
- If sidecar pages are missing or still in flight, the unsafe suffix should be recomputed normally instead of restored from L2.

Possible fix directions:

1. Track per-host-node `PoolTransferResult` / sidecar prefix metadata from L3 prefetch and constrain `load_back` by `min(kv_prefix, sidecar_prefix)`.
2. Alternatively, avoid inserting partial KV host nodes for required-sidecar hybrid pools unless their sidecar prefix is available.
3. `PoolTransferResult.update_extra_pool_hit_pages()` should use contiguous-prefix semantics for `ALL_PAGES` sidecars; summing successful pages can be unsafe if a middle INDEXER page is missing.

### Environment

Observed in a downstream deployment based on SGLang 0.5.13-series code with GLM-5.2 DSA/NSA, TP=4, CP=4, FP8 KV cache, HiCache L3 storage, and `best_effort` prefetch policy.

The relevant community main code path was checked locally at:

```text
sgl-project/sglang main commit: 430418e218
```

I can provide more details about the repro shape if needed, but the exact storage cluster and host addresses are internal and omitted here.

### PR DESCRIPTION
**Pull request by AlejandroParedesLT**
## Motivation

Fixes #30057.

For DSA/NSA (and other hybrid) models, HiCache L3 prefetch under `best_effort`/`timeout` can terminate on a partial KV prefix while a required `ALL_PAGES` sidecar pool (e.g. the DSA/NSA `INDEXER`) was never fetched, or only partially fetched. The host radix tree previously inserted such a prefix using only the KV completion length, so a later L2 load-back could restore KV pages with missing/stale sidecar state and silently corrupt sparse-attention output.

## Modifications

- Add `clamp_prefix_to_sidecar_coverage()` (`hicache_storage.py`) and apply it in both `UnifiedRadixCache` and `HiRadixCache`'s `check_prefetch_progress`, so the published prefix is always `min(kv_tokens, sidecar_tokens)` for `ALL_PAGES` sidecars. Termination timing (`best_effort`/`timeout` terminating early) is untouched — it's intentional and stays a latency bound; only the length that gets *published* is now bounded by sidecar coverage.
- `HybridCacheController._page_transfer` now attempts the sidecar fetch for whatever KV pages actually landed instead of skipping it whenever the batch didn't fully complete, and truncates the sidecar transfer's keys/host_indices to that range.
- `PoolTransferResult.update_extra_pool_hit_pages` now counts a leading contiguous run of successes for `ALL_PAGES` pools instead of summing, so a gap in the middle of a fetch can't be over-counted as a longer safe prefix.
- Extends the TP all-reduce in both classes to keep sidecar hit-page counts consistent across ranks (previously missing entirely for `HiRadixCache`, and missing for raw sidecars like `INDEXER` — as opposed to component-attached ones like SWA/Mamba — in `UnifiedRadixCache`).

## Accuracy Tests

No GPU + DSA/NSA model + real L3 backend was available in the environment this fix was developed in, so no end-to-end accuracy test was run against a real sparse-attention model. The fix is validated at the unit/integration level instead (see below); a real DSA/NSA + L3 run to confirm the corruption reported in #30057 no longer reproduces would be valuable follow-up validation.

## Speed Tests and Profiling

N/A — this only changes which prefix length is safe to publish on a partial/best-effort completion; it doesn't change the hot path for a normal full completion.

## Testing

Added `test/registered/unit/mem_cache/test_hicache_sidecar_prefix_safety.py`:
- Unit tests for `PoolTransferResult.update_extra_pool_hit_pages`'s new contiguous-count semantics and `clamp_prefix_to_sidecar_coverage` directly.
- A regression test for `HybridCacheController._page_transfer`'s partial-completion sidecar fetch.
- Integration tests that drive `UnifiedRadixCache.check_prefetch_progress` and `HiRadixCache.check_prefetch_progress` end-to-end (with `_insert_helper_host` spied) to confirm the published prefix length is actually clamped by sidecar coverage, not just computed and discarded.

All new tests were mutation-tested: reverting each part of the fix individually causes the corresponding test to fail, confirming they exercise the actual invariant rather than passing vacuously. Ran the existing `test_unified_radix_cache_unittest.py` (1040 tests) and `test_hiradix_cache_unit.py` suites with no regressions.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (no user-facing docs affected — internal cache-controller invariant)
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). (see Accuracy Tests note above — no GPU/DSA model available to validate)
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #29800075845](https://github.com/sgl-project/sglang/actions/runs/29800075845)
Latest PR Test (Extra): :x: [Run #29800075699](https://github.com/sgl-project/sglang/actions/runs/29800075699)

### CODE REVIEW

**Comment by AlejandroParedesLT** [comment]
Hi @merrymercy, @Ying1123, @hnyls2002, can any of you please review my pull request?
Thank you

**Comment by junliu-mde** [comment]
After receiving multiple internal user reports, I applied this patch to the production cluster. ~~I haven't received any similar reports since the first day; interestingly, it seems the patch is working, will keep eyes on it.~~

(On B300, GLM 5.2, 1P1D + write-through + mooncake L3)

Update: still happens, looks this PR didn't solve the problem 100%

File: python/sglang/srt/mem_cache/hicache_storage.py [hunk 1]
```diff
@@ -131,11 +131,45 @@ def update_kv_hit_pages(self, kv_hit_pages: int) -> None:
...
+        updates = {}
+        for name, rs in results.items():
+            if hit_policies.get(name) == PoolHitPolicy.ALL_PAGES:
+                try:
+                    updates[name] = rs.index(False)
+                except ValueError:
+                    updates[name] = len(rs)
+            else:
+                updates[name] = sum(rs)
+        self.extra_pool_hit_pages.update(updates)
```
Comment (preslavle) [inline python/sglang/srt/mem_cache/hicache_storage.py]: I also came to a similar design here https://github.com/sgl-project/sglang/pull/31108. I ended up using an additional field instead of repurposing this one. Potentially cleaner?

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 2]
```diff
@@ -1505,12 +1506,23 @@ def check_prefetch_progress(self, req_id: str) -> bool:
         logger.debug(f"Prefetch {req_id} completed with {completed_tokens} tokens")
 
         min_completed_tokens = completed_tokens
-        # Synchronize workers before mutating host cache tree state.
-        completed_tokens_tensor = torch.tensor(min_completed_tokens, dtype=torch.int)
-        self._all_reduce_attn_groups(
-            completed_tokens_tensor, torch.distributed.ReduceOp.MIN
+        hit_pages = operation.pool_storage_result.extra_pool_hit_pages
```
Comment (preslavle) [inline python/sglang/srt/mem_cache/hiradix_cache.py]: This can result in attribute error?

Only the DSA hybrid op carries pool_transfers/pool_storage_result; the base MHA/MLA op (shared method) has neither, so guard as in can_terminate_prefetch and keep the plain scalar all-reduce for it.

File: python/sglang/srt/mem_cache/hybrid_cache/hybrid_cache_controller.py [hunk 3]
```diff
@@ -642,30 +642,36 @@ def _page_transfer(self, operation):
...
         super()._page_transfer(operation)
 
-        # Extra pools only after KV fully completes. If KV terminated early
-        # (IO failure, timeout, TP mismatch), skip extra IO entirely to avoid
-        # data misalignment.
+        # Fetch sidecars for whatever KV pages landed; the caller clamps the
+        # usable prefix to sidecar coverage via clamp_prefix_to_sidecar_coverage.
         kv_completed_pages = operation.completed_tokens // self.page_size
-        if operation.pool_transfers and kv_completed_pages == len(operation.hash_value):
+        if operation.pool_transfers and kv_completed_pages > 0:
```
Comment (stepinto) [inline python/sglang/srt/mem_cache/hybrid_cache/hybrid_cache_controller.py]: How about just consider **full match** of KV tokens (i.e. keep the original condition (kv_completed_pages == len(...))?  I think it would make code much cleaner.

**Review by xiezhq-hermann — APPROVED** [review]

**Comment by xiezhq-hermann** [comment]
Pretty important fix, @hzh0425 @huangtingwei9988 please help take a look too

**Comment by hzh0425** [comment]
> Pretty important fix, @hzh0425 @huangtingwei9988 please help take a look too
@xiezhq-hermann 

This fix might not be optimal yet. I'd prefer to adopt an all-or-nothing strategy, because for hybrid model, if one extra pool fails to prefetch successfully, the entire prefetch result becomes meaningless.

I'm working on a fix right now.

**Comment by hzh0425** [comment]
@AlejandroParedesLT this issue should be fixed by https://github.com/sgl-project/sglang/pull/31443

could you take a try for this patch?

**Comment by hzh0425** [comment]
https://github.com/sgl-project/sglang/pull/31443#issuecomment-5033058600

should be fixed by this pr, thanks for the reporting

i've add you as the co-author
